tardis/utilities/utils.py

```diff
@@ -66,6 +67,32 @@ def csv_parser(
             }
 
 
+KeyType = TypeVar("KeyType")
+
+
+def deep_update(
+    original_mapping: Mapping[KeyType, Any],
+    mapping_update: Mapping[KeyType, Any],
+) -> Mapping[KeyType, Any]:
+    from collections.abc import Mapping
+
+    updated_mapping = deepcopy(original_mapping)
+    for key, value in mapping_update.items():
+        if isinstance(value, Mapping):
+            updated_mapping[key] = deep_update(value, original_mapping.get(key, {}))
+        elif isinstance(value, List):
+            updated_mapping[key] = []
+            for item in original_mapping.get(key, []):
+                # do not allow duplicate entries in list, so only append entries
+                # not to update
+                if item not in value:
+                    updated_mapping[key].append(deepcopy(item))
+            updated_mapping[key].extend(deepcopy(value))
```

Comment:
It might be simpler to initialise the new list directly with the items.
```suggestion
            # do not duplicate entries in the list
            updated_mapping[key] = [
                deepcopy(item)
                for item in original_mapping.get(key, [])
                if item not in value
            ]
            updated_mapping[key].extend(deepcopy(value))
```

Reply:
FWIW, my expectation is usually that existing things are not changed. So as a result I would expect *original* items to keep their ordering instead of being pushed to the end.
```suggestion
            updated_mapping[key] = deepcopy(original_mapping.get(key, []))
            # do not duplicate entries which are already in the list
            updated_mapping[key].extend(
                deepcopy(item)
                for item in value
                if item not in updated_mapping[key]
            )
```

Reply:
I doubt there is an algorithmic solution to this. People might want to either append or prepend parameters and we cannot know that unless they tell us.

Perhaps a practical approach is to make updates append but give people *two* configuration options – one that is applied as `update(input, default)` (prepending) and one as `update(default, input)` (appending).